### DISCUSSION TITLE
Multiple React roots?

### DISCUSSION THREAD

**Discussion by oriSomething** [adobe/react-spectrum#2601] [Q&A]
I need to refactor existing application to React, which means there is going to be for some time multi roots of React (or "React islands"). I would really like to use React ARIA, but I'm not sure there won't be issues with multiple React roots. 

Is it possible to use React ARIA with multiple React roots? Something I need to consider?

**Comment by snowystinger** [comment]
I know we protect against this a little bit, specifically around id's https://github.com/adobe/react-spectrum/blob/0571f1cee4565674086ea834327d0d59c093ea1f/packages/%40react-aria/ssr/src/SSRProvider.tsx#L32
and using CSS modules so that class names don't clash.
I don't think we have actually tried it though, so you may run into some interesting cases we haven't considered. I would just say proceed with caution.

**Reply by oriSomething** [reply]
The question more about cases like [this (focus control)](https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/focus/src/FocusScope.tsx#L340), which seems to have unwanted side-effects in multi roots.

I would add another concern. I use Web Components with shadow DOM to avoid 2 frameworks fighting over DOM changes. So it means rendering React inside Shadow DOM with `<slot>` as `children`. So sometimes I can have:
Legacy framework -> React -> Legacy framework -> React

I could've hack it around with patching context, but there is no "safe" way, to detect all context in a tree to move them to another React root

**Reply by reidbarber** [reply]
Context in general can't be shared between roots, so that will definitely give you trouble. I'm not aware of this being solved at the moment unfortunately.

**Reply by snowystinger** [reply]
That said, if you're just using our hooks, you probably don't need to worry about context. Our hooks can't create context. Only components can do that.
As far as globals go, we actually do attach event listeners to the window object and we do have some code around modals that poly fill aria-modal. There's ways to restrict that to sub trees, but again, proceed with caution. See https://react-spectrum.adobe.com/react-aria/watchModals.html
